File: articles/grit-burnout-and-the-risks-of-overpersistence/r/grit-burnout-overpersistence-model.R
# Grit, Burnout, and the Risks of Overpersistence
# Synthetic-data workflow for research-method demonstration only.
# Do not use this workflow to evaluate, rank, hire, admit, discipline, or assess real people.

set.seed(42)

n <- 1000

perseverance_effort <- rnorm(n)
consistency_interests <- rnorm(n)
grit <- 0.60 * perseverance_effort + 0.40 * consistency_interests

demand_intensity <- rnorm(n)
goal_rigidity <- rnorm(n)
identity_pressure <- rnorm(n)
sunk_cost <- rnorm(n)
recovery_capacity <- rnorm(n)
social_support <- rnorm(n)
autonomy <- rnorm(n)
feedback_responsiveness <- rnorm(n)
goal_fit <- rnorm(n)

overpersistence <- (
  0.22 * grit +
  0.24 * sunk_cost +
  0.22 * identity_pressure +
  0.20 * goal_rigidity -
  0.22 * feedback_responsiveness -
  0.20 * goal_fit +
  rnorm(n)
)

burnout_risk <- (
  0.24 * demand_intensity +
  0.22 * overpersistence +
  0.18 * goal_rigidity +
  0.16 * grit -
  0.26 * recovery_capacity -
  0.20 * social_support -
  0.18 * autonomy +
  rnorm(n)
)

sustainable_persistence <- (
  0.20 * grit +
  0.24 * goal_fit +
  0.22 * feedback_responsiveness +
  0.20 * recovery_capacity +
  0.18 * social_support +
  0.18 * autonomy -
  0.24 * burnout_risk -
  0.16 * overpersistence +
  rnorm(n)
)

df <- data.frame(
  perseverance_effort,
  consistency_interests,
  grit,
  demand_intensity,
  goal_rigidity,
  identity_pressure,
  sunk_cost,
  recovery_capacity,
  social_support,
  autonomy,
  feedback_responsiveness,
  goal_fit,
  overpersistence,
  burnout_risk,
  sustainable_persistence
)

grit_median <- median(df$grit)
recovery_median <- median(df$recovery_capacity)

df$profile <- ifelse(
  df$grit >= grit_median & df$recovery_capacity >= recovery_median,
  "high_grit_high_recovery",
  ifelse(
    df$grit >= grit_median & df$recovery_capacity < recovery_median,
    "high_grit_low_recovery",
    ifelse(
      df$grit < grit_median & df$recovery_capacity >= recovery_median,
      "low_grit_high_recovery",
      "low_grit_low_recovery"
    )
  )
)

dir.create("data/processed", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)

write.csv(df, "data/processed/grit-burnout-overpersistence-modeled-data-r.csv", row.names = FALSE)

correlations <- cor(df[, c(
  "grit",
  "demand_intensity",
  "goal_rigidity",
  "identity_pressure",
  "sunk_cost",
  "recovery_capacity",
  "social_support",
  "autonomy",
  "feedback_responsiveness",
  "goal_fit",
  "overpersistence",
  "burnout_risk",
  "sustainable_persistence"
)])

write.csv(correlations, "outputs/tables/grit-burnout-overpersistence-correlations-r.csv")

profile_summary <- aggregate(
  cbind(
    sustainable_persistence,
    burnout_risk,
    overpersistence,
    grit,
    recovery_capacity,
    goal_fit,
    feedback_responsiveness,
    social_support,
    autonomy
  ) ~ profile,
  data = df,
  FUN = mean
)

write.csv(profile_summary, "outputs/tables/grit-burnout-overpersistence-profile-summary-r.csv", row.names = FALSE)

model_grit_only <- lm(burnout_risk ~ grit, data = df)

model_overpersistence <- lm(
  burnout_risk ~ grit + demand_intensity + goal_rigidity + overpersistence,
  data = df
)

model_burnout <- lm(
  burnout_risk ~ grit + demand_intensity + goal_rigidity + overpersistence +
    recovery_capacity + social_support + autonomy,
  data = df
)

model_sustainable <- lm(
  sustainable_persistence ~ grit + goal_fit + feedback_responsiveness +
    recovery_capacity + social_support + autonomy + burnout_risk +
    overpersistence,
  data = df
)

comparison <- data.frame(
  model = c(
    "grit_only_burnout",
    "demand_overpersistence_burnout",
    "contextual_burnout_model",
    "sustainable_persistence_model"
  ),
  r_squared = c(
    summary(model_grit_only)$r.squared,
    summary(model_overpersistence)$r.squared,
    summary(model_burnout)$r.squared,
    summary(model_sustainable)$r.squared
  ),
  adjusted_r_squared = c(
    summary(model_grit_only)$adj.r.squared,
    summary(model_overpersistence)$adj.r.squared,
    summary(model_burnout)$adj.r.squared,
    summary(model_sustainable)$adj.r.squared
  )
)

write.csv(comparison, "outputs/tables/grit-burnout-overpersistence-model-comparison-r.csv", row.names = FALSE)
write.csv(
  summary(model_burnout)$coefficients,
  "outputs/tables/grit-burnout-overpersistence-burnout-coefficients-r.csv"
)
write.csv(
  summary(model_sustainable)$coefficients,
  "outputs/tables/grit-burnout-overpersistence-sustainable-persistence-coefficients-r.csv"
)

print(round(correlations, 3))
print(round(profile_summary, 3))
print(round(comparison, 4))
print(round(summary(model_burnout)$coefficients, 4))
print(round(summary(model_sustainable)$coefficients, 4))

File: articles/grit-burnout-and-the-risks-of-overpersistence/python/grit_burnout_overpersistence_model.py
"""
Grit, Burnout, and the Risks of Overpersistence

Synthetic-data workflow for modeling burnout risk, overpersistence, and
sustainable persistence.

This script is for article support and research-method demonstration only.
It should not be used to assess, rank, hire, admit, discipline, or evaluate real people.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-burnout-overpersistence-modeled-data-python.csv"
TABLE_OUT = ROOT / "outputs" / "tables" / "grit-burnout-overpersistence-model-comparison-python.csv"
PROFILE_OUT = ROOT / "outputs" / "tables" / "grit-burnout-overpersistence-profile-summary-python.csv"
CORR_OUT = ROOT / "outputs" / "tables" / "grit-burnout-overpersistence-correlations-python.csv"

rng = np.random.default_rng(42)
n = 1000

perseverance_effort = rng.normal(0, 1, n)
consistency_interests = rng.normal(0, 1, n)
grit = 0.60 * perseverance_effort + 0.40 * consistency_interests

demand_intensity = rng.normal(0, 1, n)
goal_rigidity = rng.normal(0, 1, n)
identity_pressure = rng.normal(0, 1, n)
sunk_cost = rng.normal(0, 1, n)
recovery_capacity = rng.normal(0, 1, n)
social_support = rng.normal(0, 1, n)
autonomy = rng.normal(0, 1, n)
feedback_responsiveness = rng.normal(0, 1, n)
goal_fit = rng.normal(0, 1, n)

overpersistence = (
    0.22 * grit
    + 0.24 * sunk_cost
    + 0.22 * identity_pressure
    + 0.20 * goal_rigidity
    - 0.22 * feedback_responsiveness
    - 0.20 * goal_fit
    + rng.normal(0, 1, n)
)

burnout_risk = (
    0.24 * demand_intensity
    + 0.22 * overpersistence
    + 0.18 * goal_rigidity
    + 0.16 * grit
    - 0.26 * recovery_capacity
    - 0.20 * social_support
    - 0.18 * autonomy
    + rng.normal(0, 1, n)
)

sustainable_persistence = (
    0.20 * grit
    + 0.24 * goal_fit
    + 0.22 * feedback_responsiveness
    + 0.20 * recovery_capacity
    + 0.18 * social_support
    + 0.18 * autonomy
    - 0.24 * burnout_risk
    - 0.16 * overpersistence
    + rng.normal(0, 1, n)
)

df = pd.DataFrame(
    {
        "perseverance_effort": perseverance_effort,
        "consistency_interests": consistency_interests,
        "grit": grit,
        "demand_intensity": demand_intensity,
        "goal_rigidity": goal_rigidity,
        "identity_pressure": identity_pressure,
        "sunk_cost": sunk_cost,
        "recovery_capacity": recovery_capacity,
        "social_support": social_support,
        "autonomy": autonomy,
        "feedback_responsiveness": feedback_responsiveness,
        "goal_fit": goal_fit,
        "overpersistence": overpersistence,
        "burnout_risk": burnout_risk,
        "sustainable_persistence": sustainable_persistence,
    }
)

grit_median = df["grit"].median()
recovery_median = df["recovery_capacity"].median()

conditions = [
    (df["grit"] >= grit_median) & (df["recovery_capacity"] >= recovery_median),
    (df["grit"] >= grit_median) & (df["recovery_capacity"] < recovery_median),
    (df["grit"] < grit_median) & (df["recovery_capacity"] >= recovery_median),
    (df["grit"] < grit_median) & (df["recovery_capacity"] < recovery_median),
]
choices = [
    "high_grit_high_recovery",
    "high_grit_low_recovery",
    "low_grit_high_recovery",
    "low_grit_low_recovery",
]
df["profile"] = np.select(conditions, choices, default="unclassified")

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
TABLE_OUT.parent.mkdir(parents=True, exist_ok=True)
PROFILE_OUT.parent.mkdir(parents=True, exist_ok=True)
CORR_OUT.parent.mkdir(parents=True, exist_ok=True)

df.to_csv(DATA_OUT, index=False)

correlations = df[
    [
        "grit",
        "demand_intensity",
        "goal_rigidity",
        "identity_pressure",
        "sunk_cost",
        "recovery_capacity",
        "social_support",
        "autonomy",
        "feedback_responsiveness",
        "goal_fit",
        "overpersistence",
        "burnout_risk",
        "sustainable_persistence",
    ]
].corr()

correlations.to_csv(CORR_OUT)

model_grit_only = sm.OLS(
    df["burnout_risk"],
    sm.add_constant(df[["grit"]]),
).fit()

model_overpersistence = sm.OLS(
    df["burnout_risk"],
    sm.add_constant(df[["grit", "demand_intensity", "goal_rigidity", "overpersistence"]]),
).fit()

model_burnout = sm.OLS(
    df["burnout_risk"],
    sm.add_constant(
        df[
            [
                "grit",
                "demand_intensity",
                "goal_rigidity",
                "overpersistence",
                "recovery_capacity",
                "social_support",
                "autonomy",
            ]
        ]
    ),
).fit()

model_sustainable = sm.OLS(
    df["sustainable_persistence"],
    sm.add_constant(
        df[
            [
                "grit",
                "goal_fit",
                "feedback_responsiveness",
                "recovery_capacity",
                "social_support",
                "autonomy",
                "burnout_risk",
                "overpersistence",
            ]
        ]
    ),
).fit()

comparison = pd.DataFrame(
    {
        "model": [
            "grit_only_burnout",
            "demand_overpersistence_burnout",
            "contextual_burnout_model",
            "sustainable_persistence_model",
        ],
        "r_squared": [
            model_grit_only.rsquared,
            model_overpersistence.rsquared,
            model_burnout.rsquared,
            model_sustainable.rsquared,
        ],
        "adjusted_r_squared": [
            model_grit_only.rsquared_adj,
            model_overpersistence.rsquared_adj,
            model_burnout.rsquared_adj,
            model_sustainable.rsquared_adj,
        ],
    }
)

profile_summary = (
    df.groupby("profile", as_index=False)
    .agg(
        n=("profile", "size"),
        mean_sustainable_persistence=("sustainable_persistence", "mean"),
        mean_burnout_risk=("burnout_risk", "mean"),
        mean_overpersistence=("overpersistence", "mean"),
        mean_grit=("grit", "mean"),
        mean_recovery_capacity=("recovery_capacity", "mean"),
        mean_goal_fit=("goal_fit", "mean"),
        mean_feedback_responsiveness=("feedback_responsiveness", "mean"),
        mean_social_support=("social_support", "mean"),
        mean_autonomy=("autonomy", "mean"),
    )
)

comparison.to_csv(TABLE_OUT, index=False)
profile_summary.to_csv(PROFILE_OUT, index=False)

print("Saved modeled dataset:", DATA_OUT)
print("Saved correlations:", CORR_OUT)
print("Saved model comparison:", TABLE_OUT)
print("Saved profile summary:", PROFILE_OUT)
print(comparison.round(4))
print(profile_summary.round(4))
